File: ios/Plugin/ADEUMMobileCapacitorPlugin.swift
import Foundation
import ADEUMInstrumentation
import Capacitor

@objc public class ADEUMMobileCapacitorPlugin: NSObject {
    
    var reference_map = [String:Any]()
    
    @objc public func generateKey() -> String {
        return UUID().uuidString
    }
    
    @objc public func echo(_ value: String) -> String {
        print(value)
        return value
    }
    @objc public func pluginInitialize(config: ADEumAgentConfiguration) -> Void {
        ADEumInstrumentation.initWith(config)
    }
    @objc public func startTimerWithName(name: String) -> Void {
        ADEumInstrumentation.startTimer(withName: name)
    }
    @objc public func stopTimerWithName(name: String) -> Void {
        ADEumInstrumentation.stopTimer(withName: name)
    }
    @objc public func reportMetricWithName(name: String, value: Int64) -> Void {
        ADEumInstrumentation.reportMetric(withName: name, value: value)
    }
    @objc public func leaveBreadcrumb(name: String?) -> Void {
        ADEumInstrumentation.leaveBreadcrumb(name)
    }
    @objc public func setUserData(key: String, value: String) -> Void {
        ADEumInstrumentation.setUserData(key, value: value)
    }
    @objc public func removeUserData(key: String) -> Void {
        ADEumInstrumentation.removeUserData(key)
    }
    @objc public func takeScreenshot() -> Void {
        ADEumInstrumentation.takeScreenshot()
    }
    /*
     * For the plugin I do not think any ios centric beginCall makes sense
    @objc public func beginCall(receiver: Any, selector: Selector, withArguments: [Any]?) -> Any? {
        //ADEumInstrumentation.beginCall(<#T##receiver: Any##Any#>, selector: <#T##Selector#>)
        return ADEumInstrumentation.beginCall(receiver, selector: selector, withArguments: withArguments)
        
    }*/
    @objc public func beginCall(className: String, methodName: String, withArguments: [Any]) -> String {
        let call = ADEumInstrumentation.beginCall(className, methodName: methodName, withArguments: withArguments)
        let key = self.generateKey()
        reference_map[key] = call
        return key
    }
    
    @objc public func endCall(tracker_key: String) -> Void {
        let tracker = reference_map[tracker_key]
        ADEumInstrumentation.endCall(tracker)
    }
    
    @objc public func getCorrelationHeaders() -> Any {
        let headers = ADEumServerCorrelationHeaders.generate()
        return headers
    }
    
    @objc public func beginHttpRequest(url: URL) -> String {
        
        let http_tracker = ADEumHTTPRequestTracker.init(url: url)
        let key = self.generateKey()
        reference_map[key] = http_tracker
        return key
    }
    
    @objc public func reportDone(tracker_key: String) -> Void {
        
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        tracker?.reportDone()
        reference_map.removeValue(forKey: tracker_key)
    }
    
    
    @objc public func withResponseCode(tracker_key: String, statusCode: NSNumber) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        //let http_tracker = tracker as! ADEumHTTPRequestTracker
        tracker?.statusCode = statusCode
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func withResponseContentLength(tracker_key: String, responseContentLength: NSNumber) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        let tmpHeaders = tracker?.allHeaderFields as! NSMutableDictionary
        tmpHeaders.setValue(responseContentLength, forKeyPath: "Content-Length")
        tracker?.allHeaderFields = tmpHeaders as? [AnyHashable : Any]
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func withRequestContentLength(tracker_key: String, requestContentLength: NSNumber) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        let tmpHeaders = tracker?.allRequestHeaderFields as! NSMutableDictionary
        tmpHeaders.setValue(requestContentLength, forKeyPath: "Content-Length")
        tracker?.allRequestHeaderFields = tmpHeaders as? [AnyHashable : Any]
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func withResponseHeaderFields(tracker_key: String, responseHeaders: NSDictionary) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        tracker?.allHeaderFields = responseHeaders as? [AnyHashable : Any]
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func withRequestHeaderFields(tracker_key: String, requestHeaders: NSDictionary) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        tracker?.allRequestHeaderFields = requestHeaders as? [AnyHashable : Any]
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func withInstrumentationSource(tracker_key: String, informationSource: String) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        tracker?.instrumentationSource = informationSource
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func withErrorMessage(tracker_key: String, errorMessage: String) {
        let tracker = reference_map[tracker_key] as? ADEumHTTPRequestTracker
        let http_error = NSError(domain: "com.appdynamics.ionic", code: -100, userInfo: [NSLocalizedDescriptionKey: errorMessage])
        tracker?.error = http_error
        reference_map.updateValue(tracker as Any, forKey: tracker_key)
    }
    
    @objc public func startNextSession() -> Void {
        ADEumInstrumentation.startNextSession()
    }
    
    @objc public func unblockScreenshots() -> Void {
        ADEumInstrumentation.unblockScreenshots()
    }
    
    @objc public func blockScreenshots() -> Void {
        ADEumInstrumentation.blockScreenshots()
    }
    
    @objc public func screenshotsBlocked() -> Bool {
        return ADEumInstrumentation.screenshotsBlocked()
    }
    
    @objc public func startSessionFrame(sessionFrameName: String) -> String {
        let session = ADEumInstrumentation.startSessionFrame(sessionFrameName)
        let key = self.generateKey()
        reference_map[key] = session
        return key
    }
    
    @objc public func endSessionFrame(session_key: String) -> Void {
        let session = reference_map[session_key] as? ADEumSessionFrame
        session?.end()
        reference_map.removeValue(forKey: session_key)
    }
    
    @objc public func updateSessionFrameName(session_key: String, session_name: String) -> Void {
        let session = reference_map[session_key] as? ADEumSessionFrame
        session?.updateName(session_name)
        reference_map.updateValue(session as Any, forKey: session_key)
    }
    
    @objc public func getVersion() -> String {
        var version = "1.0.0"
        version = String(ADEumInstrumentation.version())
        return version
    }
    
    @objc public func clear() -> Void {
        reference_map.removeAll()
    }
    
    /*
     - (void)updateSessionFrameName:(CDVInvokedUrlCommand*)command {
         CDVPluginResult *result;
         @try {
             if ([command.arguments count] < 2) {
                 [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
                 NSLog(@"Obj-C updateSessionFrameName called -- bad command");
             } else if ([command.arguments count] >= 2) {
                 NSString *key = [command argumentAtIndex:0 withDefault:nil andClass:[NSString class]];
                 NSString *sessionFrameName = [command argumentAtIndex:1 withDefault:nil andClass:[NSString class]];
                 if ([key length] > 0 && [sessionFrameName length] > 0) {
                     ADEumSessionFrame *sessionFrame = [self.keymap objectForKey:key];
                     if (sessionFrame != nil) {
                         [sessionFrame updateName:sessionFrameName];
                         NSLog(@"updateSessionFrameName %@, %@", key, sessionFrameName);
                         result = [self finalizeSuccessfulResult:nil];
                     } else {
                         result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"SessionFrame not found"];
                     }
                 } else {
                     result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:@"Invalid SessionFrame object"];
                 }
             }
         } @catch (NSException *exception) {
             result = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:[exception reason]];
         }
         [self.commandDelegate sendPluginResult:result callbackId:command.callbackId];
     }
     */
}

File: android/src/main/java/com/appdynamics/appd/mrum/plugins/ionic/ADEUMMobileCapacitorPlugin.java
package com.appdynamics.appd.mrum.plugins.ionic;

import android.util.Log;
//import com.appdynamics.eumagent.runtime.Instrumentation;
import com.appdynamics.eumagent.runtime.AgentConfiguration;
import com.appdynamics.eumagent.runtime.CallTracker;
import com.appdynamics.eumagent.runtime.HttpRequestTracker;
import com.appdynamics.eumagent.runtime.SessionFrame;
import com.appdynamics.eumagent.runtime.ServerCorrelationHeaders;
import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;


import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.HashMap;
import java.util.ArrayList;
import java.util.Iterator;
import java.net.URL;
import java.net.MalformedURLException;

public class ADEUMMobileCapacitorPlugin {
    private final static String ADEumPluginType = "Ionic Capacitor";
    private final static String VERSION = "1.1.0";
    private final static String TAG = "ADEUMCapacitorPlgImpl";
    private HashMap<String, CallTracker> callTrackers = new HashMap<>();
    private HashMap<String, HttpRequestTracker> httpRequestTrackers = new HashMap<>();
    private HashMap<String, SessionFrame> sessionFrames = new HashMap<>();
    public boolean pluginInitialized;

    public ADEUMMobileCapacitorPlugin() {
    }



    public void crash() throws RuntimeException {
        throw new RuntimeException("Crash Attempt");
    }

    public void flush() {
        try {
            HttpRequestTracker tracker = com.appdynamics.eumagent.runtime.Instrumentation.beginHttpRequest(new URL("http://flush.queue"));
            tracker.withResponseCode(200).reportDone();
            //callbackContext.success();
        } catch (MalformedURLException mue) {
        }
    }

    public boolean checkPluginInitialized() {
        return pluginInitialized;
    }

    /* Custom method implementation starts here */
    public String echo(String value) {
        Log.i("Echo", value);
        return value;
    }

    public void pluginInitialize(AgentConfiguration config){
        com.appdynamics.eumagent.runtime.Instrumentation.start(config);
        pluginInitialized = true;
    }

    public void startTimerWithName(String name) {
        com.appdynamics.eumagent.runtime.Instrumentation.startTimer(name);
    }


    public void stopTimerWithName(String name) {
        com.appdynamics.eumagent.runtime.Instrumentation.stopTimer(name);
    }

    public void reportMetricWithName(String name, Integer value) {
        com.appdynamics.eumagent.runtime.Instrumentation.reportMetric(name, value);
    }

    public void leaveBreadcrumb(String name) {
        com.appdynamics.eumagent.runtime.Instrumentation.leaveBreadcrumb(name);
    }

    public void setUserData(String key, String value) {
        com.appdynamics.eumagent.runtime.Instrumentation.setUserData(key, value);
    }

    public void removeUserData(String key) {
        //function is missing from sdk
        //com.appdynamics.eumagent.runtime.Instrumentation.removeUserData(key);
    }

    public void takeScreenshot() {
        com.appdynamics.eumagent.runtime.Instrumentation.takeScreenshot();
    }

    public String beginCall(String className, String methodName, JSArray args) {
        CallTracker callTracker = com.appdynamics.eumagent.runtime.Instrumentation.beginCall(className, methodName, args);
        String uuid = UUID.randomUUID().toString();
        callTrackers.put(uuid, callTracker);
        return uuid;
    }

    public void endCall(JSObject callTrackerId) {
        CallTracker callTracker = callTrackers.get(callTrackerId);
        callTracker.reportCallEnded();
        callTrackers.remove(callTrackerId);
    }

    public String beginHttpRequest(String url) {
        String uuid = UUID.randomUUID().toString();
        try {
            HttpRequestTracker tracker = com.appdynamics.eumagent.runtime.Instrumentation.beginHttpRequest(new URL(url));
            httpRequestTrackers.put(uuid, tracker);
        } catch (MalformedURLException me){
            Log.e(TAG, me.getMessage());
        }
        return uuid;
    }

    public void reportDone(String trackerId) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            tracker.reportDone();
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public void withResponseCode(String trackerId, Integer statusCode) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            tracker.withResponseCode(statusCode);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public void withResponseContentLength(String trackerId, long contentLength) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            tracker.withResponseContentLength(contentLength);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public void withRequestContentLength(String trackerId, long contentLength) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            tracker.withResponseContentLength(contentLength);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public void withResponseHeaderFields(String trackerId, JSObject httpHeaders) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            HashMap headersMap = new HashMap();
            Iterator itor = httpHeaders.keys();
            while (itor.hasNext()) {
                String key = (String) itor.next();
                String val = httpHeaders.getString(key);
                ArrayList list = new ArrayList();
                list.add(val);
                // AppD magic headers must be uppercase CORE-39486
                if (key.startsWith("adrum")) {
                    key = key.toUpperCase();
                }
                headersMap.put(key, list);
            }
            tracker.withResponseHeaderFields(headersMap);
        }catch(ClassCastException | NullPointerException e){
            Log.e(TAG, "Invalid Tracker Object");
        }
        catch(Exception e){
            Log.e(TAG, "Internal error occurred setting header fields");
        }
    }

    public void withRequestHeaderFields(String trackerId, JSObject httpHeaders) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            HashMap headersMap = new HashMap();
            Iterator itor = httpHeaders.keys();
            while (itor.hasNext()) {
                String key = (String) itor.next();
                String val = httpHeaders.getString(key);
                ArrayList list = new ArrayList();
                list.add(val);
                // AppD magic headers must be uppercase CORE-39486
                if (key.startsWith("adrum")) {
                    key = key.toUpperCase();
                }
                headersMap.put(key, list);
            }
            tracker.withRequestHeaderFields(headersMap);
        }catch(ClassCastException | NullPointerException e){
            Log.e(TAG, "Invalid Tracker Object");
        }
        catch(Exception e){
            Log.e(TAG, "Internal error occurred setting header fields");
        }
    }

    public void withInstrumentationSource(String trackerId, String information_source) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            tracker.withInstrumentationSource(information_source);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public void withErrorMessage(String trackerId, String errorMessage) {
        try{
            HttpRequestTracker tracker = httpRequestTrackers.get(trackerId);
            tracker.withError(errorMessage);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public Map<String, List<String>> getCorrelationHeaders() {
        Map<String, List<String>> correlationHeaders = ServerCorrelationHeaders.generate();
        return correlationHeaders;
    }

    public void startNextSession() {
        com.appdynamics.eumagent.runtime.Instrumentation.startNextSession();
    }

    public void unblockScreenshots() {
        com.appdynamics.eumagent.runtime.Instrumentation.unblockScreenshots();
    }

    public void blockScreenshots() {
        com.appdynamics.eumagent.runtime.Instrumentation.blockScreenshots();
    }

    public Boolean screenshotsBlocked() {
        return com.appdynamics.eumagent.runtime.Instrumentation.screenshotsBlocked();
    }

    public String startSessionFrame(String sessionFrameName) {
        SessionFrame sessionFrame = com.appdynamics.eumagent.runtime.Instrumentation.startSessionFrame(sessionFrameName);
        String uuid = UUID.randomUUID().toString();
        sessionFrames.put(uuid, sessionFrame);
        return uuid;
    }

    public void endSessionFrame(String session_key) {
        try{
            SessionFrame sessionFrame = sessionFrames.get(session_key);
            sessionFrame.end();
            sessionFrames.remove(session_key);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public void updateSessionFrame(String session_key, String sessionFrameName) {
        try{
            SessionFrame sessionFrame = sessionFrames.get(session_key);
            sessionFrame.updateName(sessionFrameName);
            sessionFrames.put(session_key, sessionFrame);
        }catch(Exception e){
            Log.e(TAG, "Invalid Tracker Object");
        }
    }

    public String getVersion() {
        return VERSION;
    }

    public void clear() {
        if (callTrackers != null)
            callTrackers.clear();
        if (httpRequestTrackers != null)
            httpRequestTrackers.clear();
        if (sessionFrames != null)
            sessionFrames.clear();
    }




}
